sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala

```diff
@@ -249,9 +249,9 @@ case class Cbrt(child: Expression) extends UnaryMathExpression(math.cbrt, "CBRT"
   """,
   since = "1.4.0",
   group = "math_funcs")
-case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL") {
+ case class Ceil(child: Expression) extends UnaryMathExpression(math.ceil, "CEIL") {
   override def dataType: DataType = child.dataType match {
```

Comment:
Let's define the return type we want:
1. if input is decimal type, we should follow `RoundBase` to define the return type
2. if input is integral type, returning `LongType` as before should be good.
3. if input is float/double, returning `LongType` is definitely wrong. I think returning `DoubleType` should be good.

The problem is what to do if the `scala` parameter is not given. Shall we keep backward compatibility and use the same return type as before? Or we prefer consistency within the system and change the return type?

It looks weird if `ceil(c_double)` and `ceil(c_double, 0)` have different data type. Another idea is to only accept integer constant as the scale parameter, then we can make `ceil` return long type for float/double input if `scale <= 0`.

cc @maropu @viirya @srielau

Reply:
@cloud-fan thanks for your review. In the current implementation, if the `scale` parameter is 0 or not specified, we follow the default behaviour of the `Ceil`, in which for the input type float/double, we return `LongType`. If the `scale` parameter is specified and greater than 0, we follow the `RoundBase` behaviour.  I believe `ceil(c_double)` and `ceil(c_double, 0)` will have the same behaviour.

Reply:
OK let's only accept int constants for the scale parameter, so that we can keep the return type for `scale <= 0`. We probably need to implement `ExpressionBuilder`:
```
object CeilExpressionBuilder extends ExpressionBuilder {
  def build(expressions: Seq[Expression]): Expression = {
    if (expressions.length == 1) Ceil(expressions.head, 0, LongType)
    else if (expressions.length == 2 && !expressions.last.foldable) {
       fail...
    } else if ...
  }
}

case class Ceil(child: Expression, scale: Int, dataType: DataType)
```

Reply:
Do you think will it possibly that `scale` to be not only a literal in the future? Currently looks like the way to keep `ceil(c_double)` and `ceil(c_double, 0)` have consistent data type is to check `scale` value before constructing `Ceil`/`RoundCeil` expr. Seems if `scale` isn't a literal, we cannot do this. If we are pretty sure `scale` cannot be non-literal in the future, then it is fine.

Reply:
If we need to support non-literal scale parameter in the future, we just make this behavior change and let `ceil(c_coule, scale)` always return double. I don't see an issue with delaying this behavior change, or never make this behavior change if literal scale parameter is fine.

Reply:
@cloud-fan You are right, we need `ExpressionBuilder` for the sql part. I've added the expression builder for `ceil` and `floor` and enough test cases in `MathFunctionsSuite` and it looks fine. I'll push the modifications along with other review comments.

```
object CeilExpressionBuilder extends ExpressionBuilder {
  def build(expressions: Seq[Expression]): Expression = {
    if (expressions.length == 1) Ceil(expressions.head)
    else if (expressions.length == 2) Ceil(expressions(0), expressions(1))
    else throw new AnalysisException("Function ceil cannot take more than 2 parameters")
  }
}

object FloorExpressionBuilder extends ExpressionBuilder {
  def build(expressions: Seq[Expression]): Expression = {
    if (expressions.length == 1) Floor(expressions.head)
    else if (expressions.length == 2) Floor(expressions(0), expressions(1))
    else throw new AnalysisException("Function floor cannot take more than 2 parameters")
  }
}

```

And I believe, for now it will be safer to conserve the same behaviour for `ceil`/`floor` without `scale` parameter or when `scale = 0` and follow the behaviour of `round` when scale parameter is specified and `>0`. I think you prefer the same ?

Reply:
We have 2 choices:
1. return double if scale > 0 and return long if scale <= 0. Only accept int literals as the scale parameter, null is also forbidden.
2. always return double, and return null if scale is null.

Can we pick option 1 first as it's stricter? We can always go to option 2 later if it's necessary.

Reply:
I agree, for scale <=0 we can always return Long, but for scale > 0 we always return double even when the input type is `Long`/`Int` ? It's better to follow the same behaviour as `round` returning the same output type as the input type for scale > 0, no ?

Reply:
Sorry I was not very clear. I was talking about `ceil` function with float/double input. For integral input, it should definitely return long type always.

Reply:
@cloud-fan I think we can do something like this, what do you think ?

```
object Ceil {
  def apply(child: Expression, scale: Expression): Expression = {
    val scaleV = scale.eval(EmptyRow)
    if (scaleV == null) return null

    (child.dataType, scaleV.asInstanceOf[Int]) match {
      case (_, 0) => Ceil(child)
      case (_, x) if x < 0 => RoundCeil(Cast(child, LongType), scale)
      case (_: IntegralType | ShortType, _) => RoundCeil(Cast(child, LongType), scale)
      case (_ : FloatType, _) => RoundCeil(Cast(child, DoubleType), scale)
      case _ => RoundCeil(child, scale)
    }
  }
}
```

Reply:
Could we can do the same using `DecimalType.bounded(..)`/ `DecimalType.Fixed(..)` ?

Reply:
> if (scaleV == null) return null

We can't do it because we need to return an expression which must define a data type.

Reply:
I thought the same, but the `eval` of RoundBase does it.

```
override def eval(input: InternalRow): Any = {
    if (scaleV == null) { // if scale is null, no need to eval its child at all
      null
    } else {
      val evalE = child.eval(input)
      if (evalE == null) {
        null
      } else {
        nullSafeEval(evalE)
      }
    }
  }
```

So if we must define the data type of the return expression, we could do something like : 
```
if (scaleV == null) return Literal.create(null, dtNullScale(child.dataType))

def dtNullScale(dt: DataType): DataType = dt match {
  case _ : FloatType | _ : DoubleType => DoubleType
  case _ => LongType
}
```

What do you think ?

Reply:
It doesn't matter how we implement it. The problem is we don't have a clear answer for the return type of `ceil(c_double, null)`. `ceil(c_double, 1)` returns double, `ceil(c_double, -1)` returns long, I don't think either long or double makes sense for `ceil(c_double, null)`.

Reply:
Or we could let `RoundBase` to manage null value for scale parameter as it does for `round`/`bround`.

Reply:
For me personally, returning `DoubleType` for `ceil(c_double, null)` don't shock me, but returning `LongType` for `ceil(c_double, null)` might probably do. And the same goes for `ceil(c_long, null)`.

But for the moment, i think we could either 
1. Let `RoundBase` to deal with the scale will null value. or
2. Throw `AnalysisException` if the scale is null. something like : 
`if (scaleV == null) throw new AnalysisException("Scale parameter can not be null")`

Reply:
I'd prefer option 2. This allows us to change the behavior later without a breaking change.

Reply:
I agree with you. I'm adding few unit tests to verify the schema of the RoundCeil and RoundFloor. Is there any other tests or changes we may need to add to complete this feature ? I'll push the changes all together. Thanks !